tests/math/sequence/test_sequential_funcs.py

```diff
@@ -141,6 +171,81 @@ def test_mean_bkwd():
     pass
 
 
+@fwdprop_test_factory(
+    mygrad_func=average, true_func=np.average, num_arrays=1, kwargs=gen_average_args,
+)
+def test_average_fwd():
+    pass
+
+
+def test_average_weights_zero_sum():
+    with raises(ZeroDivisionError):
+        average(np.array([1, 2]), weights=np.array([-1, 1]))
+
+
+def test_average_scl():
+    x = mg.Tensor(np.array([1, 2, 3, 4]))
+    weights = mg.Tensor(np.array([1, 2, 3, 4]))
+    avg, scl = average(x, weights=weights, returned=True)
+    expected_avg, expected_scl = np.average(x.data, weights=weights.data, returned=True)
+    assert np.isclose(avg.data, expected_avg)
+    assert np.isclose(scl.data, expected_scl)
+    scl.backward()
+    assert avg.grad is None
+    assert weights.grad is not None
+
+
+def test_canonicalize_weights():
+    x = np.empty((2, 2))
+    # Same shape, no-op
+    assert _canonicalize_weights(x, axis=0, weights=np.empty((2, 2))).shape == x.shape
+    # Differing shapes, make weights broadcastable
+    assert _canonicalize_weights(x, axis=0, weights=np.empty((2,))).shape == (2, 1)
+    assert _canonicalize_weights(x, axis=1, weights=np.empty((2,))).shape == (1, 2)
+
+
+def test_canonicalize_weights_misuse():
+    x = np.empty((2, 2))
+    with raises(TypeError):
+        # Missing axis
+        _canonicalize_weights(x, axis=None, weights=np.empty((2,)))
+    with raises(TypeError):
+        # Weights not 1D
+        _canonicalize_weights(x, axis=0, weights=np.empty((2, 1)))
+    with raises(ValueError):
+        # Incompatible weights length
+        _canonicalize_weights(x, axis=0, weights=np.empty((3,)))
+
+
+@backprop_test_factory(
+    mygrad_func=average,
+    true_func=np.average,
+    num_arrays=1,
+    kwargs=gen_average_args,
+    index_to_bnds={0: (-10, 10)},
+    vary_each_element=True,
+)
+def test_average_bkwd():
+    pass
+
+
+# Helps test the 2nd Tensor output of average.
+def average_returned_wrapper(average_func):
+    return lambda x, axis, weights: average_func(x, axis=axis, weights=weights, returned=True)[1]
+
+
+@backprop_test_factory(
+    mygrad_func=average_returned_wrapper(average),
+    true_func=average_returned_wrapper(np.average),
+    num_arrays=1,
+    kwargs=gen_average_args,
+    index_to_bnds={0: (-10, 10)},
+    vary_each_element=True,
+)
+def test_average_scl_bkwd():
```

Comment:
@rsokl I have added a test for `scl` which is awkward because gradient of `scl` with respect to the input is 0. So the numerical derivative is returning 0 while `x.grad` is returning `None`, after `scl.backward()` is run, because `x` is not part of `scl`'s computational graph. 

If we want, one way to handle this is to `assert_allclose` between `np.broadcast_to(0, d_num.shape)` and `d_num` when `arr.grad` is `None` at https://github.com/bantmen/MyGrad/blob/c6d6d878596aba65e98ea129d280485e67111c01/tests/wrappers/uber.py#L632